File: platforms/android/java/app/msime/client/ReturnKeyAction.java
package app.msime.client;

/** Android editor-action labels shared by return-key rendering and dispatch. */
public final class ReturnKeyAction {
    private static final int UNSPECIFIED = 0;
    private static final int NONE = 1;
    private static final int GO = 2;
    private static final int SEARCH = 3;
    private static final int SEND = 4;
    private static final int NEXT = 5;
    private static final int DONE = 6;
    private static final int PREVIOUS = 7;

    private ReturnKeyAction() {}

    public static boolean performsEditorAction(int action, boolean disabled) {
        if (disabled) return false;
        return switch (action) {
            case GO, SEARCH, SEND, NEXT, DONE, PREVIOUS -> true;
            default -> false;
        };
    }

    /** A handled composition consumes Return before any editor action or newline is considered. */
    public static boolean shouldPerformEditorAction(
            int action, boolean disabled, boolean compositionHandled) {
        return !compositionHandled && performsEditorAction(action, disabled);
    }

    public static String title(int action, boolean disabled) {
        if (!performsEditorAction(action, disabled)) return "换行";
        return switch (action) {
            case GO -> "前往";
            case SEARCH -> "搜索";
            case SEND -> "发送";
            case NEXT -> "下一项";
            case DONE -> "完成";
            case PREVIOUS -> "上一项";
            default -> "换行";
        };
    }
}

File: platforms/harmony/entry/src/main/ets/keyboard/input/ReturnKeyAction.ts
/**
 * Editor-action labels shared by return-key rendering and dispatch, ported from
 * platforms/android/java/app/msime/client/ReturnKeyAction.java.
 *
 * The numbers are the editor action constants the host receives; they are kept rather than renamed so
 * the mapping stays checkable against what the framework sends.
 */
const GO: number = 2;
const SEARCH: number = 3;
const SEND: number = 4;
const NEXT: number = 5;
const DONE: number = 6;
const PREVIOUS: number = 7;

export enum ReturnDispatch {
  EDITOR = 'editor',
  FINISH_COMPOSITION = 'finish-composition',
  COMMIT_HIGHLIGHTED = 'commit-highlighted',
  COMMIT_READING = 'commit-reading'
}

export class ReturnKeyAction {
  static dispatch(japanese: boolean, composing: boolean, candidateCount: number): ReturnDispatch {
    if (japanese && composing) {
      return candidateCount > 0 ? ReturnDispatch.COMMIT_HIGHLIGHTED : ReturnDispatch.COMMIT_READING;
    }
    if (candidateCount > 0) {
      return ReturnDispatch.COMMIT_HIGHLIGHTED;
    }
    return composing ? ReturnDispatch.FINISH_COMPOSITION : ReturnDispatch.EDITOR;
  }

  static performsEditorAction(action: number, disabled: boolean): boolean {
    if (disabled) {
      return false;
    }
    return action === GO || action === SEARCH || action === SEND
      || action === NEXT || action === DONE || action === PREVIOUS;
  }

  /** A handled composition consumes Return before any editor action or newline is considered. */
  static shouldPerformEditorAction(action: number, disabled: boolean,
                                   compositionHandled: boolean): boolean {
    return !compositionHandled && ReturnKeyAction.performsEditorAction(action, disabled);
  }

  static title(action: number, disabled: boolean): string {
    if (!ReturnKeyAction.performsEditorAction(action, disabled)) {
      return '换行';
    }
    switch (action) {
      case GO: return '前往';
      case SEARCH: return '搜索';
      case SEND: return '发送';
      case NEXT: return '下一项';
      case DONE: return '完成';
      case PREVIOUS: return '上一项';
      default: return '换行';
    }
  }
}
